File: src/erlang/ch01/1.3/l0/src/remainder.erl
-module(remainder).
-export([div_identity/2, week_shift/2, powmod/3]).

%% Return {Q, R} s.t. N = D*Q + R (like Python divmod)
div_identity(N, D) when D =:= 0 ->
    erlang:error(badarg);
div_identity(N, D) ->
    {N div D, N rem D}.

%% Wrap days (0..6, 0=Mon); robust for negatives
week_shift(Start, Shift) ->
    M = 7,
    T = ((Start rem M) + M) rem M,
    S = ((Shift rem M) + M) rem M,
    (T + S) rem M.

%% Binary exponentiation: (Base^Exp) mod Mod
powmod(_Base, _Exp, 1) -> 0;
powmod(Base, Exp, Mod) ->
    powmod_( ((Base rem Mod) + Mod) rem Mod
           , Exp
           , Mod
           , 1 rem Mod ).

powmod_(_B, 0, _M, Acc) -> Acc;
powmod_(B, E, M, Acc) when (E band 1) =:= 1 ->
    powmod_((B*B) rem M, E bsr 1, M, (Acc*B) rem M);
powmod_(B, E, M, Acc) ->
    powmod_((B*B) rem M, E bsr 1, M, Acc).

File: archived/v1/src/c/ch01/1.3/l0/remainder.c
#include "remainder.h"
#include <stdbool.h>

void div_identity(int64_t n, int64_t d, int64_t *q, int64_t *r) {
    /* caller must ensure d != 0 */
    *q = n / d;
    *r = n % d;
}

int week_shift(int start, int shift) {
    int m = 7;
    int t = (start % m + m) % m;
    int s = (shift % m + m) % m;
    return (t + s) % m;
}

int64_t powmod(int64_t base, int64_t exp, int64_t mod) {
    if (mod == 1) return 0;
    int64_t res = 1 % mod;
    int64_t b = ((base % mod) + mod) % mod;
    int64_t e = exp;
    while (e > 0) {
        if (e & 1) res = (res * b) % mod;
        b = (b * b) % mod;
        e >>= 1;
    }
    return res;
}
